common/src/main/java/io/pravega/common/io/FileModificationWatcher.java

```diff
@@ -0,0 +1,195 @@
+     *                          is useful for testing only.
+     * @throws InvalidPathException if {@code fileToWatch} is invalid
+     * @throws FileNotFoundException when a file at specified path {@code fileToWatch} does not exist
+     * @throws FileNotFoundException when a file at specified path {@code fileToWatch} does not exist
+     */
+    @VisibleForTesting
+    FileModificationWatcher(@NonNull Path pathOfFileToWatch, @NonNull Consumer<WatchEvent<?>> callback,
+                                   boolean loopContinuously)
+            throws FileNotFoundException {
+        super();
+
+        this.pathOfFileToWatch = pathOfFileToWatch;
+
+        if (!this.pathOfFileToWatch.toFile().exists()) {
+            throw new FileNotFoundException(String.format("File [%s] does not exist.", pathOfFileToWatch));
+        }
+        this.callback = callback;
+        setUncaughtExceptionHandler(uncaughtExceptionalHandler);
+
+        // Set the name for this object/thread for identification purposes.
+        this.setName("file-update-watcher-" + THREAD_NUM.incrementAndGet());
+        this.loopContinuously = loopContinuously;
+    }
+
+    @VisibleForTesting
+    String nameOfFileToWatch() {
+        Path fileName = this.pathOfFileToWatch.getFileName();
+        if (fileName != null) {
+            return fileName.toString();
+        } else {
+            throw new RuntimeException("File name is null");
+        }
+    }
+
+    @VisibleForTesting
+    Path directoryOfFileToWatch() {
+        assert this.pathOfFileToWatch != null;
+        return this.pathOfFileToWatch.getParent();
+    }
+
+    @Override
+    @SuppressWarnings("SleepWhileInLoop")
+    public void run() {
+        WatchKey watchKey = null;
+        WatchService watchService = null;
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            log.debug("Done creating watch service for watching file at path: {}", this.pathOfFileToWatch);
+
+            String fileName = nameOfFileToWatch();
+            Path directoryPath = directoryOfFileToWatch();
+            log.debug("Directory being watched is {}", directoryPath);
+
+            assert directoryPath != null;
+            directoryPath.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY,
+                    StandardWatchEventKinds.ENTRY_CREATE);
+            log.debug("Done setting up watch for modify entries for file at path: {}", this.pathOfFileToWatch);
+
+            while (true) {
+                watchKey = watchService.take();
+                log.info("Retrieved and removed watch key for watching file at path: {}", this.pathOfFileToWatch);
+
+                // Looks odd, right? Using the logic or de-duplicating file change events, as suggested by some here:
+                // https://stackoverflow.com/questions/16777869/java-7-watchservice-ignoring-multiple-occurrences-of-
+                // the-same-event
+                Thread.sleep(200);
```

Comment:
Rather than having a thread per-file where each thread blocks in a loop. It would be better to have a pool where each task is run with a fixed delay. Then there can be fewer threads than files and while one is blocking another file can be checked.

See https://guava.dev/releases/snapshot/api/docs/com/google/common/util/concurrent/AbstractScheduledService.html
which provides an easy API to do this.

Reply:
A task pool might also make it easier to reschedule the reload in the edge case of an IOException when reading the cert file due to a race.   I honestly think that file change notification may arrive while the file is still open for write.

Reply:
Currently, the `FileModificationWatcher` is used to monitor only a single file - TLS certificate file for Segment Store. Granted, it is a generic class that can find a use for other things as well, and that the change you are suggesting would be very useful. But, I think it might be prudent to enhance it further if and when we really need it. 

Would you agree?

Reply:
*If* we're sure there is only ever going to be one of these in a given process. A dedicated thread and a sleep is probably fine. Are we sure of that?

Reply:
As of now, it is going to be monitoring a single file, so there is going to be one of it in a given process. Things might change in the future, and we might have to enhance it then.